ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryUrlSigningService.java

```diff
@@ -0,0 +1,138 @@
+import java.util.Map;
+
+
+/**
+ * Default implementation of {@link UrlSigningService} that currently only converts all POST headers into query
+ * parameters and does not actually reliably sign the URL.
+ * <p/>
+ * A real implementation needs to encode enough information for manipulations to be discovered.
+ */
+public class AmbryUrlSigningService implements UrlSigningService {
+  static final String AMBRY_PARAMETERS_PREFIX = "x-ambry-";
+  private static final String LINK_EXPIRY_TIME = "et";
+  private static final String ENDPOINT_SUFFIX = "/";
+  private static final String QUERY_STRING_START = "?";
+  private static final String PARAMETER_SEPARATOR = "&";
+  private static final String PARAMETER_ASSIGN = "=";
+
+  private final String uploadEndpoint;
+  private final String downloadEndpoint;
+  private final long defaultUrlTtlSecs;
+  private final long defaultMaxUploadSize;
+  private final long maxUrlTtlSecs;
+  private final Time time;
+
+  AmbryUrlSigningService(String uploadEndpoint, String downloadEndpoint, long defaultUrlTtlSecs,
+      long defaultMaxUploadSize, long maxUrlTtlSecs, Time time) {
+    if (!uploadEndpoint.endsWith(ENDPOINT_SUFFIX)) {
+      uploadEndpoint = uploadEndpoint + ENDPOINT_SUFFIX;
+    }
+    if (!downloadEndpoint.endsWith(ENDPOINT_SUFFIX)) {
+      downloadEndpoint = downloadEndpoint + ENDPOINT_SUFFIX;
+    }
+    this.uploadEndpoint = uploadEndpoint;
+    this.downloadEndpoint = downloadEndpoint;
+    this.defaultUrlTtlSecs = defaultUrlTtlSecs;
+    this.defaultMaxUploadSize = defaultMaxUploadSize;
+    this.maxUrlTtlSecs = maxUrlTtlSecs;
+    this.time = time;
+  }
+
+  @Override
+  public String getSignedUrl(RestRequest restRequest) throws RestServiceException {
+    Map<String, Object> args = restRequest.getArgs();
+    String restMethodInSignedUrlStr = RestUtils.getHeader(args, RestUtils.Headers.URL_TYPE, true);
+    RestMethod restMethodInSignedUrl;
+    try {
+      restMethodInSignedUrl = RestMethod.valueOf(restMethodInSignedUrlStr);
+    } catch (IllegalArgumentException e) {
+      throw new RestServiceException("Unrecognized RestMethod: " + restMethodInSignedUrlStr,
+          RestServiceErrorCode.InvalidArgs);
+    }
+    StringBuilder urlBuilder = new StringBuilder();
+    switch (restMethodInSignedUrl) {
+      case GET:
+        urlBuilder.append(downloadEndpoint).append(QUERY_STRING_START);
+        break;
+      case POST:
+        urlBuilder.append(uploadEndpoint).append(QUERY_STRING_START);
+        break;
+      default:
+        throw new RestServiceException("Signing request for " + restMethodInSignedUrl + " is not supported",
+            RestServiceErrorCode.InvalidArgs);
+    }
+    long urlTtlSecs = defaultUrlTtlSecs;
+    long maxUploadSize = defaultMaxUploadSize;
+    for (Map.Entry<String, Object> entry : args.entrySet()) {
+      String name = entry.getKey();
+      Object value = entry.getValue();
+      if (name.startsWith(AMBRY_PARAMETERS_PREFIX) && value instanceof String) {
```

Comment:
Have the args been lowercased already? are query params/headers case sensitive?

Reply:
Good point - now using a different function